python/ql/test/experimental/dataflow/coverage/classes.py

```diff
@@ -57,216 +57,216 @@ def __init__(self):
 
 
 def test_init():
-    with_init = With_init()
+    with_init = With_init()  #$ MISSING: arg1="SSA variable with_init" func=With_init.__init__
 
 
 # object.__del__(self)
 class With_del:
     def __del__(self):
-        SINK1(self)  # Flow not found
+        SINK1(self)
         OK()  # Call not found
 
 
 def test_del():
-    with_del = With_del()
+    with_del = With_del()  #$ MISSING: arg1="SSA variable with_del" func=With_del.__del__
```

Comment:
Shouldn't the annotation be on the next line, where the actual `del` operation happens? (same for many other cases in this file)

Reply:
he annotation is on the source of the flow. That will not always be an argument to a call, but that seems natural enough to me?

Reply:
Well, just doing `with_del = With_del()` doesn't cause any flow to the `__del__` method. Only when you do `del with_del` will there be flow, so that's why I'm thinking that the annotation should also be on that line.

For example, if you do more than one call, with the current annotation setup you would get them mixed

```py
example_obj = ExampleClass() # $ arg1="SSA variable example_obj" func=ExampleClass.__repr__ func=ExampleClass.__del__
repr(example_obj)
del example_obj
```

and you wouldn't be able to see whether each call made it to the right function (just that both functions were reached).

I think it's more natural to get the annotations as

```py
example_obj = ExampleClass()
repr(example_obj) # $ arg1="SSA variable example_obj" func=ExampleClass.__repr__ 
del example_obj # $ arg1="SSA variable example_obj" func=ExampleClass.__del__ 
```

Now there might be something in the test setup that means this wouldn't be easy to do?

Reply:
Yeah, the test setup does not look at calls and callables at all, it just relies on sources and sinks being relevant. Perhaps if we had sources be arguments, that would be enough. That might also find arguments to special methods in the expected way.

We could also have parameters as sinks rather than having the methods sink everything. Then the test-files might shrink a few hundred lines..

Reply:
On the other hand, there is something clean about not having the test configuration use the machinery for detecting calls and callables. The tests are currently written to test a single routing each, and it seems prudent enough to keep doing that? If we want to test more complicated combinations, we could create a different configuration and setup for that...